ts/Stock/Indicators/DMI/DMIIndicator.ts

```diff
@@ -0,0 +1,398 @@
+        },
+        /**
+         * +DI line options.
+         */
+        plusDILine: {
+            /**
+             * Styles for the +DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#00ff00'
+            }
+        },
+        /**
+         * -DI line options.
+         */
+        minusDILine: {
+            /**
+             * Styles for the -DI line.
+             */
+            styles: {
+                /**
+                 * Pixel width of the line.
+                 */
+                lineWidth: 1,
+                /**
+                 * Color of the line.
+                 *
+                 * @type {Highcharts.ColorString}
+                 */
+                lineColor: '#ff0000'
+            }
+        },
+        dataGrouping: {
+            approximation: 'averages'
+        }
+    } as DMIOptions);
+
+    /* *
+     *
+     *  Functions
+     *
+     * */
+
+    public calculateDM(
+        yVal: Array<Array<number>>,
+        i: number,
+        isPositiveDM?: boolean
+    ): number {
+        const currentHigh = yVal[i][1],
+            currentLow = yVal[i][2],
+            previousHigh = yVal[i - 1][1],
+            previousLow = yVal[i - 1][2];
+
+        // If any value is null, return 0.
+        if (!currentHigh || !currentLow || !previousHigh || !previousLow) {
+            return 0;
+        }
+
+        let DM: number = 0;
+
+        if (currentHigh - previousHigh > previousLow - currentLow) {
+            // for +DM
+            DM = isPositiveDM ? Math.max(currentHigh - previousHigh, 0) : 0;
+        } else {
+            // for -DM
+            DM = !isPositiveDM ? Math.max(previousLow - currentLow, 0) : 0;
+        }
+
+        return correctFloat(DM);
+    }
+
+    public calculateDI(
+        smoothedDM: number,
+        tr: number
+    ): number {
+        return smoothedDM / (tr || 1) * 100;
+    }
+
+    public calculateDX(
+        plusDI: number,
+        minusDI: number
+    ): number {
+        return correctFloat(
+            Math.abs(plusDI - minusDI) / (Math.abs(plusDI + minusDI) || 1) * 100
+        );
+    }
+
+    public smoothValues(
+        accumulatedValues: number,
+        currentValue: number,
+        period: number
+    ): number {
+        return correctFloat(
+            accumulatedValues - accumulatedValues / period + currentValue
+        );
+    }
+
+    public getTR(
+        currentPoint: Array<number>,
+        prevPoint?: Array<number>
+    ): number {
+        // If any value is null, return 0.
+        if (!currentPoint[1] || !currentPoint[2] || !currentPoint[3]) {
+            return 0;
+        }
+
+        const HL = currentPoint[1] - currentPoint[2],
```

Comment:
You are not using these variables again, I would make it in one line with some additional comments what is what. 
Like:
```
return correctFloat(
   Math.max(
       currentPoint[1] - currentPoint[2], // high - low
       ....
   );

)

Reply:
Hmm, I'm gonna defend this. Yes, initially I had it written the way you are proposing, but, in this case (4 variables each used twice) it is much more readable with properly named and defined variables.

However, if others share your opinion, I will change it.
@pawellysy @pawelfus what do you think?